File: producer.py
import time
import sys
import json
from pyndn import Name
from pyndn import Data
from pyndn import Face
from pyndn.security import KeyChain
from pyndn.meta_info import MetaInfo
from pyndn.util import MemoryContentCache

def dump(*list):
    result = ""
    for element in list:
        result += (element if type(element) is str else str(element)) + " "
    print(result)

def main():
    face = Face()
    keyChain = KeyChain()
    face.setCommandSigningInfo(keyChain, keyChain.getDefaultCertificateName())
    memCache = MemoryContentCache(face)

    def onRegisterFailed(prefix):
        dump("Register failed for prefix", prefix.toUri())

    prefix = Name("/ndn/repo/case/test")

    dump("Register prefix", prefix.toUri())
    memCache.registerPrefix(prefix, onRegisterFailed, None, memCache.getStorePendingInterest())

    latestFrameNo = 0
    def generateFrame():
        data = Data(Name(prefix).append('s').append(Name.Component.fromSequenceNumber(latestFrameNo)))
        content = 'frame-data'
        data.setContent(content)
        metaInfo = MetaInfo()
        metaInfo.setFreshnessPeriod(30)
        data.setMetaInfo(metaInfo)
        keyChain.sign(data)
        dump("published frame", data.getName(), content)
        memCache.add(data)

    metaVersion = 0
    def updateMeta():
        data = Data(Name(prefix).append('_meta').append(Name.Component.fromVersion(metaVersion)))
        content = json.dumps({'timestamp':time.time(), 'stream':'s'})
        data.setContent(content)
        metaInfo = MetaInfo()
        metaInfo.setFreshnessPeriod(30)
        data.setMetaInfo(metaInfo)
        keyChain.sign(data)
        dump("update meta", data.getName(), content)
        memCache.add(data)

    def updateLatest(fNo):
        # data = Data(Name(prefix).append('_latest').append(Name.Component.fromVersion(metaVersion)))
        data = Data(Name(prefix).append('s').append('_latest').append(Name.Component.fromTimestamp(time.time()*1000)))
        content = Name(prefix).append('s').append(Name.Component.fromSequenceNumber(latestFrameNo)).toUri()
        data.setContent(content)
        metaInfo = MetaInfo()
        metaInfo.setFreshnessPeriod(10)
        data.setMetaInfo(metaInfo)
        keyChain.sign(data)
        dump("update _latest", data.getName(), content)
        memCache.add(data)

    metaUpdateInterval = 1
    frameGenerationInterval = 1/25
    latestUpdateInterval = 1/100
    timestampCheck = { 'meta':time.time(), 'frame_generate': time.time(), 'latest_ptr': time.time() }

    updateMeta()
    generateFrame()

    while True:
        now = time.time()

        if now - timestampCheck['meta'] >= metaUpdateInterval:
            timestampCheck['meta'] = now
            metaVersion += 1
            updateMeta()

        if now - timestampCheck['frame_generate'] >= frameGenerationInterval:
            timestampCheck['frame_generate'] = now
            latestFrameNo += 1
            generateFrame()

        if now - timestampCheck['latest_ptr'] >= latestUpdateInterval:
            timestampCheck['latest_ptr'] = now
            updateLatest(latestFrameNo)

        face.processEvents()
        time.sleep(0.001)

    face.shutdown()

main()

File: cpp/producer.cpp
// g++ -o producer producer.cpp -std=c++11 -I../../ndnrtc-thirdparty/ndn-cpp/build/include -L../../ndnrtc-thirdparty/ndn-cpp/build/lib -lndn-cpp

#include <iostream>
#include <chrono>
#include <unistd.h>
#include <ndn-cpp/name.hpp>
#include <ndn-cpp/data.hpp>
#include <ndn-cpp/interest.hpp>
#include <ndn-cpp/security/key-chain.hpp>
#include <ndn-cpp/util/memory-content-cache.hpp>

using namespace std;
using namespace std::chrono;
using namespace ndn;

int main()
{
  Face face = Face();
  KeyChain keyChain = KeyChain();
  face.setCommandSigningInfo(keyChain, keyChain.getDefaultCertificateName());
  MemoryContentCache memCache = MemoryContentCache(&face);
  Name prefix("/ndn/repo/case/test");

  cout << "register prefix " << prefix << endl;
  memCache.registerPrefix(prefix,
    [](const ptr_lib::shared_ptr<const Name>& prefix){
      cout << "register failed" << endl;
    },
    memCache.getStorePendingInterest());

  int latestFrameNo = 0;
  function<void()> generateFrame = [prefix, &latestFrameNo, &keyChain, &memCache](){
    Data data(Name(prefix).append("s").appendSequenceNumber(latestFrameNo));
    data.setContent(Blob::fromRawStr("frame-data"));
    data.getMetaInfo().setFreshnessPeriod(30);
    keyChain.sign(data);

    cout << "published frame " << data.getName() << endl;
    memCache.add(data);
  };

  int metaVersion = 0;
  function<void()> updateMeta = [prefix, &metaVersion, &keyChain, &memCache](){
    int64_t now = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
    stringstream ss;
    ss << "s:" << now;

    Data data(Name(prefix).append("_meta").appendVersion(metaVersion));
    data.setContent(Blob::fromRawStr(ss.str()));
    data.getMetaInfo().setFreshnessPeriod(30);
    keyChain.sign(data);

    cout << "update _meta " << data.getName() << endl;
    memCache.add(data);
  };

  function<void()> updateLatest = [prefix, &latestFrameNo, &keyChain, &memCache](){
    int64_t now = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
    Data data(Name(prefix).append("s").append("_latest").appendTimestamp(now));
    data.setContent(Blob::fromRawStr(Name(prefix).append("s").appendSequenceNumber(latestFrameNo).toUri()));
    data.getMetaInfo().setFreshnessPeriod(10);
    keyChain.sign(data);

    cout << "update _latest " << data.getName() << endl;
    memCache.add(data);
  };

  double metaUpdateInterval = 1000;
  double frameGenerationInterval = 33;
  double latestUpdateInterval = 10;

  int64_t now = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
  map<string, int64_t> timestampCheck {{"meta", now},{"frame_generate", now}, {"latest_ptr", now}};

  updateMeta();
  generateFrame();

  while (true) {
    now = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();

    if (now - timestampCheck["meta"] >= metaUpdateInterval)
    {
      timestampCheck["meta"] = now;
      metaVersion += 1;
      updateMeta();
    }

    if (now - timestampCheck["frame_generate"] >= frameGenerationInterval)
    {
      timestampCheck["frame_generate"] = now;
      latestFrameNo += 1;
      generateFrame();
    }

    if (now - timestampCheck["latest_ptr"] >= latestUpdateInterval)
    {
      timestampCheck["latest_ptr"] = now;
      updateLatest();
    }

    face.processEvents();
    usleep(10);
  }
}
